### PR TITLE
Improve best batch and random forest classes

### PR DESCRIPTION
**Pull request by AldoGl**
## Proposed changes

I generalise the best batch constructor by letting it take the `a` and `b` parameter of the beta-binomial distribution as well as the `perturbation_range` applied to the parameter values. 

I generalise the random forest constructor by letting it take the number of classes (`n_classes`) used by the classifier in the fitting phase.

### CODE REVIEW

File: black_it/samplers/best_batch.py [hunk 1]
```diff
@@ -25,7 +27,41 @@
...
+            max_deduplication_passes: the maximum number of deduplication passes that are made
+            a: the a parameter of the beta-binomial distribution
+            b: the b parameter of the beta-binomial distribution
+            perturbation_range: the range of the perturbation applied. The actual perturbation will be in the range
+                plus/minus the perturbation_range times the precision of the specific parameter coordinate
+        """
+        super().__init__(batch_size, random_state, max_deduplication_passes)
+        self.a = a
+        self.b = b
+        self.perturbation_range = perturbation_range
```
Comment (marcofavoritobi) [inline black_it/samplers/best_batch.py:64]: It is OK for me. To make the new feature perfect, I would add some value checks on `a`, `b`, and `perturbation_range`. In particular:

- `a > 0.0`
- `b > 0.0`
- `perturbation_range > 1`

Comment (AldoGl) [reply black_it/samplers/best_batch.py:64]: Thanks, it makes sense. Probablu `perturbation_range` should be greater or equal to 1

Comment (marcofavoritobi) [reply black_it/samplers/best_batch.py:64]: > Thanks, it makes sense. Probablu `perturbation_range` should be greater or equal to 1

If `perturbation_range = 1`, `integers(1, self.perturbation_range)` gives error:

```
np.random.default_rng().integers(1, 1) # raises ValueError: low >= high
```

**Review by marcofavoritobi — APPROVED** [review]

File: black_it/samplers/best_batch.py (hunk 1 above)
Comment (AldoGl) [reply black_it/samplers/best_batch.py:78]: thanks!

File: black_it/samplers/random_forest.py [hunk 2]
```diff
@@ -30,32 +30,36 @@
...
-            criterion: The function to measure the quality of a split.
+            criterion: the function to measure the quality of a split.
+            n_classes: the number of classes used in the random forest. The classes are selected as the quantiles
+                of the distribution of loss values.
         """
         super().__init__(batch_size, random_state, max_deduplication_passes)
 
         self._n_estimators = n_estimators
         self._criterion = criterion
+        self._n_classes = n_classes
```
[the remark this answers is not in the document]
Comment (AldoGl) [reply black_it/samplers/random_forest.py:67]: thanks!

**Comment by AldoGl** [comment]
Thank you @marcofavoritobi for your feedback. I implemented the checks you mentioned

**Pull request merged**